benchmark/C/Savina/src/parallelism/GuidedSearch.lf

```diff
@@ -0,0 +1,246 @@
+ * Akka implementation, each node is send as an individual message. However, in
+ * LF, we cannot send multiple messages at the same tag, and introducing a delay
+ * via a logical action would be very costly for many messages. Instead, we send
+ * only a single message containing a list of nodes.
+ *
+ * It is also important to note that the workers in this benchmark operate on
+ * shared state (the grid). Each node has an attribute that indicates its
+ * parent. Initially, this attribute is uninitialized, but it is updated when it
+ * is visited by a worker. Thus, the parent attribute can also be used to
+ * determine whether a node was already visited. However, multiple reactions may
+ * access the parent attribute of the same node at the same time. Thus,
+ * similarly to the Akka implementation, the attribute is implemented as an
+ * atomic variable. However, this does not fully eliminate the race condition.
+ * Sometimes one reaction might write first and sometimes another reaction
+ * writes first, effectively introducing nondeterminism into the benchmark.
+ *
+ * The peculiarities of this design also lead to another quirk. The benchmark
+ * does not scale well with the number of threads. A single threaded execution
+ * is much faster than a multi threaded one. Interestingly the same effect can
+ * be observed in the Akka implementation. This indicates that the problem lies
+ * in the benchmark itself, and not (only) in the LF runtime. In fact,
+ * multi-threaded executions have the tendency to perform much more work as
+ * compared to single threaded execution. This is likely due to the race-conditions
+ * mentioned before.
+ *  
+ * @author Christian Menard
+ * @author Hannes Klein
+ * @author Soroush Bateni
+ */
+
+target CCpp {
+    /* [[[cog
+      if (threaded_runtime=="True"):
+          cog.outl(f"threads: {threads},")
+      else:
+          cog.outl("threads: 0,")
+    ]]] */
+    threads: 0,
+    /// [[[end]]]
+    build-type : RelWithDebInfo,
+    cmake-include: "GuidedSearch/GuidedSearch.cmake",
+    files: ["GuidedSearch/GuidedSearchCommon.cc", 
+        "GuidedSearch/GuidedSearchCommon.hh", 
+        "GuidedSearch/GridNode.cc", 
+        "GuidedSearch/GridNode.hh"],
+    logging: "warn"
+};
+
+preamble {=
+    #include "GuidedSearchCommon.hh"
+    #include <deque>
+    #include <list>
+    #include <iostream>
+=}
+
+reactor Manager(numWorkers: size_t(20), gridSize: size_t(30), priorities: size_t(30)) {
+
+    state grid: Grid*;
+    state nodesToVisit: std::deque<GridNode*>;
+    state pathWasFound: bool(false);
+    
+    input start: bool;
+    output finished: bool;
+    
+    output[numWorkers] search: WorkMessage;
+    input[numWorkers] pathFound: bool;
+    input[numWorkers] moreNodesToVisit: std::deque<GridNode*>;
+    
+    logical action next;
+    
+    reaction (start) -> next {=
+        self->grid = new Grid(self->gridSize, self->priorities);
```

Comment:
this should be freed somewhere. Actually, why is this a pointer? Couldn't this be a simple state variable? Since this is the CCpp target, you could also consider using a smart pointer to free the data automagically.